Duplicati/Library/Main/Operation/Backup/BackendUploader.cs

```diff
@@ -85,98 +122,323 @@ public static Task Run(Common.BackendHandler backend, Options options, Common.Da
+                        if (worker == null)
                         {
-                            lastSize = ((VolumeUploadRequest)req).BlockVolume.SourceSize;
+                            worker = new Worker(m_backendFactory());
+                            workers.Add(worker);
+                        }
 
-                            if (noIndexFiles || ((VolumeUploadRequest)req).IndexVolume == null)
-                                task = new KeyValuePair<int, Task>(1, backend.UploadFileAsync(((VolumeUploadRequest)req).BlockVolume, null));
+                        if (req is VolumeUploadRequest volumeUpload)
+                        {
+                            if (volumeUpload.IndexVolume == null)
+                                worker.Task = Task.Run(() => UploadFileAsync(volumeUpload.BlockEntry, worker, m_cancelTokenSource.Token));
                             else
-                                task = new KeyValuePair<int, Task>(2, backend.UploadFileAsync(((VolumeUploadRequest)req).BlockVolume, name => ((VolumeUploadRequest)req).IndexVolume.CreateVolume(name, options, database)));
+                                worker.Task = Task.Run(() => UploadBlockAndIndexAsync(volumeUpload, worker, m_cancelTokenSource.Token));
+
+                            lastSize = volumeUpload.BlockVolume.SourceSize;
+                            uploadsInProgress++;
+                        }
+                        else if (req is FilesetUploadRequest filesetUpload)
+                        {
+                            worker.Task = Task.Run(() => UploadVolumeWriter(filesetUpload.Fileset, worker, m_cancelTokenSource.Token));
+                            uploadsInProgress++;
+                        }
+                        else if (req is IndexVolumeUploadRequest indexUpload)
+                        {
+                            worker.Task = Task.Run(() => UploadVolumeWriter(indexUpload.IndexVolume, worker, m_cancelTokenSource.Token));
+                            uploadsInProgress++;
                         }
-                        else if (req is FilesetUploadRequest)
-                            task = new KeyValuePair<int, Task>(1, backend.UploadFileAsync(((FilesetUploadRequest)req).Fileset));
-                        else if (req is IndexVolumeUploadRequest)
-                            task = new KeyValuePair<int, Task>(1, backend.UploadFileAsync(((IndexVolumeUploadRequest)req).IndexVolume));
-                        else if (req is FlushRequest)
+                        else if (req is FlushRequest flush)
                         {
                             try
                             {
-                                while(inProgress.Count > 0)
-                                    await inProgress.Dequeue().Value;
-                                active = 0;
+                                await Task.WhenAll(workers.Select(w => w.Task));

```

Comment:
I'm not sure if I'm reading this correctly, but it seems that previously if a queued `Task` threw an exception, the remaining `Tasks` would not be awaited.  Now, all the `Tasks` are run until they complete.  If that's true, what are the side effects (if any) or differences in behavior (good or bad)?

Reply:
I think you're reading that right. I was in the mindset of completing the uploads so it seemed to make sense to use `WhenAll` but you bring up a good point. In normal operation if an exception occurs we want to cancel all of the other currently in progress uploads so we should have the same behavior here where we are waiting for the last few uploads.

I think the downside to using `WhenAll` here would be that if an exception occurs we would wait for the other files to finish or throw their own exceptions, which could potentially take a long time. And then later on those files would just get deleted from the backend anyway.

In normal operation where everything works properly this is obviously the nicest way to go about waiting for the uploads. Unless you have other thoughts I'll go ahead with changing this to cancel the other uploads when an exception occurs.